### PR TITLE
Fix K-Caching Issues

### PR DESCRIPTION
**Pull request by fthaler**
## Description

Fixes two issues of k-Caches as reported by @huanglangwen:
- Limit k-caches to vertical loop sections with a single horizontal execution.
- Fix possible symbol name duplication during the splitting of a vertical loop section when local scalars are present.

## Requirements

Before submitting this PR, please make sure:

- [x] The code builds cleanly without new errors or warnings
- [x] The code passes all the existing tests
- [x] If this PR adds a new feature, new tests have been added to test these
new features
- [x] All relevant documentation has been updated or added

Additionally, if this PR contains code authored by new contributors:

- [x] All the authors are covered by a valid contributor assignment agreement,
signed by the employer if needed, provided to ETH Zurich
- [x] The names of all the new contributors have been added to an updated
version of the AUTHORS.rst file included in the PR

### CODE REVIEW

File: src/gtc/passes/oir_optimizations/caches.py [hunk 1]
```diff
@@ -105,7 +105,9 @@ class KCacheDetection(NodeTranslator):
     max_cacheable_offset: int = 5
 
     def visit_VerticalLoop(self, node: oir.VerticalLoop, **kwargs: Any) -> oir.VerticalLoop:
-        if node.loop_order == common.LoopOrder.PARALLEL:
+        if node.loop_order == common.LoopOrder.PARALLEL or any(
```
Comment (havogt) [inline src/gtc/passes/oir_optimizations/caches.py:108]: Please add a comment why limiting to a single horizontal_execution is a requirement. Is it because of the kcache living in the scope of the horizontal execution?

File: tests/test_unittest/test_gtc/test_passes/test_oir_optimizations/test_caches.py [hunk 2]
```diff
@@ -382,13 +383,18 @@ def test_fill_to_local_k_caches_section_splitting_forward():
...
+                            HorizontalExecutionFactory(
+                                body=[
+                                    AssignStmtFactory(
+                                        left__name="foo", right__name="foo", right__offset__k=0
+                                    ),
+                                    AssignStmtFactory(
+                                        left__name="foo", right__name="foo", right__offset__k=1
+                                    ),
+                                ],
+                                declarations=[LocalScalarFactory()],
```
Comment (havogt) [inline tests/test_unittest/test_gtc/test_passes/test_oir_optimizations/test_caches.py:396]: Is this change related to the name clash? I cannot see why...

File: tests/test_unittest/test_gtc/test_passes/test_oir_optimizations/test_caches.py [hunk 3]
```diff
@@ -382,13 +383,18 @@ def test_fill_to_local_k_caches_section_splitting_forward():
...
-                                left__name="foo", right__name="foo", right__offset__k=1
-                            ),
+                        horizontal_executions=[
+                            HorizontalExecutionFactory(
+                                body=[
+                                    AssignStmtFactory(
+                                        left__name="foo", right__name="foo", right__offset__k=0
+                                    ),
+                                    AssignStmtFactory(
+                                        left__name="foo", right__name="foo", right__offset__k=1
```
Comment (havogt) [inline tests/test_unittest/test_gtc/test_passes/test_oir_optimizations/test_caches.py:393]: I am wondering if this is a good test. It has a race condition, so if we would run a checker on this piece it would trigger. On the other hand, if we force ourselves to always write "correct" snippets for unittest it could result in more complicated test setups and distraction from the actual codepath that is tested. Opinions?

**Review by havogt — CHANGES_REQUESTED** [review]

File: src/gtc/passes/oir_optimizations/caches.py (hunk 1 above)
Comment (fthaler) [reply src/gtc/passes/oir_optimizations/caches.py:108]: Yes, exactly. Usually this restriction should not matter at all, as all horizontal executions without horizontal offsets should be merged. Currently though this is not possible as soon as there are masks, because there are no if statements yet.

File: tests/test_unittest/test_gtc/test_passes/test_oir_optimizations/test_caches.py (hunk 2 above)
Comment (fthaler) [reply tests/test_unittest/test_gtc/test_passes/test_oir_optimizations/test_caches.py:396]: Yes it is. The local scalar gets duplicated when the section is split. So we have a name clash if we do not rename one of the local scalars.

File: tests/test_unittest/test_gtc/test_passes/test_oir_optimizations/test_caches.py (hunk 3 above)
Comment (fthaler) [reply tests/test_unittest/test_gtc/test_passes/test_oir_optimizations/test_caches.py:393]: I can’t see a race condition here, it’s a forward loop. Sure it does not make any computational sense to just overwrite foo here, but it is just about the read and write offsets. So this is probably the simplest code to get a write to `foo` and a read from `foo` and `foo[0, 0, 1]`.

Comment (havogt) [reply tests/test_unittest/test_gtc/test_passes/test_oir_optimizations/test_caches.py:393]: Sorry, I "saw" a horizontal offset.

File: tests/test_unittest/test_gtc/test_passes/test_oir_optimizations/test_caches.py (hunk 2 above)
Comment (havogt) [reply tests/test_unittest/test_gtc/test_passes/test_oir_optimizations/test_caches.py:396]: makes sense.

File: src/gtc/passes/oir_optimizations/caches.py (hunk 1 above)
Comment (havogt) [reply src/gtc/passes/oir_optimizations/caches.py:108]: Please add this info as a comment.

**Review by havogt — APPROVED** [review]

File: src/gtc/passes/oir_optimizations/caches.py (hunk 1 above)
Comment (fthaler) [reply src/gtc/passes/oir_optimizations/caches.py:111]: Done.

**Comment by havogt** [comment]
bors try
